main.tf

```diff
@@ -2,11 +2,34 @@
 # This file is used to implement the ROOT module
 *********************************************************************/
 locals {
-  kp_backup_crn        = var.backup_encryption_key_crn != null ? var.backup_encryption_key_crn : var.key_protect_key_crn
+  # The backup encryption key crn doesn't support Hyper Protect Crypto Service (HPCS) at the moment. If 'backup_encryption_key_crn' is null, will use 'kms_key_crn' as encryption key if its Key Protect key otherwise it will use using randomly generated keys.
+  # https://cloud.ibm.com/docs/cloud-databases?topic=cloud-databases-hpcs&interface=cli
+  kp_backup_crn = var.backup_encryption_key_crn != null ? var.backup_encryption_key_crn : (can(regex(".*kms.*", var.kms_key_crn)) ? var.kms_key_crn : null)
+
   auto_scaling_enabled = var.auto_scaling == null ? [] : [1]
+
+  kms_service = var.kms_key_crn != null ? (
+    can(regex(".*kms.*", var.kms_key_crn)) ? "kms" : (
+      can(regex(".*hs-crypto.*", var.kms_key_crn)) ? "hs-crypto" : null
+    )
+  ) : null
```

Comment:
See recent changes to this code in 34eb271c47f1f75600e6b692b1e53d20aa80120d in postgresql module.

With existing validation it is possible to get to this code and generate null, which causes the subsequent kms_policy to fail.